File: ts/linkself/test/roster.test.ts
import { describe, expect, it } from "vitest";
import { generateIdentity } from "../src/did.js";
import {
  buildRoster,
  marshalRoster,
  rosterDeviceDIDs,
  rosterHasDevice,
  unmarshalRoster,
  verifyRoster,
  withDevice,
  withoutDevice,
  type DeviceEntry,
} from "../src/roster.js";

// Helper: a device DID is just another Ed25519 did:key (the device's transport key).
async function deviceEntry(label: string): Promise<DeviceEntry> {
  const id = await generateIdentity();
  return { deviceDID: id.did, label };
}

describe("device roster (signed by the user key)", () => {
  it("builds a roster the user key can verify", async () => {
    const user = await generateIdentity();
    const d1 = await deviceEntry("PC");
    const d2 = await deviceEntry("Phone");
    const roster = await buildRoster(user, [d1, d2]);

    expect(roster.userDID).toBe(user.did);
    expect(rosterDeviceDIDs(roster).sort()).toEqual(
      [d1.deviceDID, d2.deviceDID].sort(),
    );
    expect(await verifyRoster(roster)).toBe(true);
  });

  it("is order-independent (canonical encoding sorts devices)", async () => {
    const user = await generateIdentity();
    const d1 = await deviceEntry("A");
    const d2 = await deviceEntry("B");
    const r1 = await buildRoster(user, [d1, d2]);
    const r2 = await buildRoster(user, [d2, d1]);
    expect(r1.sig).toEqual(r2.sig);
  });

  it("de-duplicates devices by deviceDID (last wins)", async () => {
    const user = await generateIdentity();
    const d = await deviceEntry("old");
    const roster = await buildRoster(user, [d, { ...d, label: "new" }]);
    expect(roster.devices).toHaveLength(1);
    expect(roster.devices[0]!.label).toBe("new");
    expect(await verifyRoster(roster)).toBe(true);
  });

  it("rejects a roster signed by a different (impostor) key", async () => {
    const user = await generateIdentity();
    const impostor = await generateIdentity();
    const d1 = await deviceEntry("PC");
    const good = await buildRoster(user, [d1]);
    // Keep the signature but claim a different userDID ⇒ verify fails.
    const forged = { ...good, userDID: impostor.did };
    expect(await verifyRoster(forged)).toBe(false);
  });

  it("rejects a tampered device list", async () => {
    const user = await generateIdentity();
    const d1 = await deviceEntry("PC");
    const roster = await buildRoster(user, [d1]);
    const injected = await deviceEntry("rogue");
    const tampered = { ...roster, devices: [...roster.devices, injected] };
    expect(await verifyRoster(tampered)).toBe(false);
  });

  it("adds and removes devices, re-signing each time", async () => {
    const user = await generateIdentity();
    const d1 = await deviceEntry("PC");
    const d2 = await deviceEntry("Phone");

    let roster = await buildRoster(user, [d1]);
    roster = await withDevice(user, roster.devices, d2);
    expect(rosterHasDevice(roster, d2.deviceDID)).toBe(true);
    expect(await verifyRoster(roster)).toBe(true);

    roster = await withoutDevice(user, roster.devices, d1.deviceDID);
    expect(rosterHasDevice(roster, d1.deviceDID)).toBe(false);
    expect(rosterHasDevice(roster, d2.deviceDID)).toBe(true);
    expect(await verifyRoster(roster)).toBe(true);
  });

  it("round-trips through marshal/unmarshal", async () => {
    const user = await generateIdentity();
    const d1 = await deviceEntry("PC");
    const d2 = await deviceEntry("Phone");
    const roster = await buildRoster(user, [d1, d2]);

    const restored = unmarshalRoster(marshalRoster(roster));
    expect(restored.userDID).toBe(roster.userDID);
    expect(restored.devices).toEqual(roster.devices);
    // Byte equality (sign() may return a Buffer; unmarshal a plain Uint8Array).
    expect(Array.from(restored.sig)).toEqual(Array.from(roster.sig));
    expect(await verifyRoster(restored)).toBe(true);
  });
});

File: core/internal/roster/roster_test.go
package roster

import (
	"testing"

	"github.com/SeijiShii/link-self/core/internal/did"
)

func deviceEntry(t *testing.T, label string) DeviceEntry {
	t.Helper()
	id, err := did.Generate()
	if err != nil {
		t.Fatalf("generate device identity: %v", err)
	}
	return DeviceEntry{DeviceDID: id.DID, Label: label}
}

func TestBuildAndVerify(t *testing.T) {
	user, err := did.Generate()
	if err != nil {
		t.Fatal(err)
	}
	d1 := deviceEntry(t, "PC")
	d2 := deviceEntry(t, "Phone")
	r, err := Build(user, []DeviceEntry{d1, d2})
	if err != nil {
		t.Fatal(err)
	}
	if r.UserDID != user.DID {
		t.Errorf("UserDID = %q, want %q", r.UserDID, user.DID)
	}
	if !HasDevice(r, d1.DeviceDID) || !HasDevice(r, d2.DeviceDID) {
		t.Error("roster missing a device")
	}
	if !Verify(r) {
		t.Error("Verify() = false, want true")
	}
}

func TestSignatureIsOrderIndependent(t *testing.T) {
	user, _ := did.Generate()
	d1 := deviceEntry(t, "A")
	d2 := deviceEntry(t, "B")
	r1, _ := Build(user, []DeviceEntry{d1, d2})
	r2, _ := Build(user, []DeviceEntry{d2, d1})
	if string(r1.Sig) != string(r2.Sig) {
		t.Error("signatures differ for reordered devices")
	}
}

func TestDedupeLastWins(t *testing.T) {
	user, _ := did.Generate()
	d := deviceEntry(t, "old")
	r, err := Build(user, []DeviceEntry{d, {DeviceDID: d.DeviceDID, Label: "new"}})
	if err != nil {
		t.Fatal(err)
	}
	if len(r.Devices) != 1 {
		t.Fatalf("len(Devices) = %d, want 1", len(r.Devices))
	}
	if r.Devices[0].Label != "new" {
		t.Errorf("Label = %q, want new", r.Devices[0].Label)
	}
	if !Verify(r) {
		t.Error("Verify() = false")
	}
}

func TestRejectImpostorUserDID(t *testing.T) {
	user, _ := did.Generate()
	impostor, _ := did.Generate()
	r, _ := Build(user, []DeviceEntry{deviceEntry(t, "PC")})
	r.UserDID = impostor.DID // keep signature, claim a different user
	if Verify(r) {
		t.Error("Verify() = true for impostor userDID, want false")
	}
}

func TestRejectTamperedDevices(t *testing.T) {
	user, _ := did.Generate()
	r, _ := Build(user, []DeviceEntry{deviceEntry(t, "PC")})
	r.Devices = append(r.Devices, deviceEntry(t, "rogue"))
	if Verify(r) {
		t.Error("Verify() = true for tampered device list, want false")
	}
}

func TestWithAndWithoutDevice(t *testing.T) {
	user, _ := did.Generate()
	d1 := deviceEntry(t, "PC")
	d2 := deviceEntry(t, "Phone")

	r, _ := Build(user, []DeviceEntry{d1})
	r, err := WithDevice(user, r.Devices, d2)
	if err != nil {
		t.Fatal(err)
	}
	if !HasDevice(r, d2.DeviceDID) || !Verify(r) {
		t.Error("WithDevice failed")
	}

	r, err = WithoutDevice(user, r.Devices, d1.DeviceDID)
	if err != nil {
		t.Fatal(err)
	}
	if HasDevice(r, d1.DeviceDID) {
		t.Error("device not removed")
	}
	if !HasDevice(r, d2.DeviceDID) || !Verify(r) {
		t.Error("WithoutDevice failed")
	}
}

func TestMarshalRoundTrip(t *testing.T) {
	user, _ := did.Generate()
	r, _ := Build(user, []DeviceEntry{deviceEntry(t, "PC"), deviceEntry(t, "Phone")})
	data, err := Marshal(r)
	if err != nil {
		t.Fatal(err)
	}
	got, err := Unmarshal(data)
	if err != nil {
		t.Fatal(err)
	}
	if got.UserDID != r.UserDID || len(got.Devices) != len(r.Devices) {
		t.Error("round-trip mismatch")
	}
	if !Verify(got) {
		t.Error("Verify() = false after round-trip")
	}
}
